src/include/common/item_pointer.h

```diff
@@ -54,20 +57,30 @@ class ItemPointerComparator {
     return (p1->block == p2->block) && (p1->offset == p2->offset);
   }
 
+  bool operator()(ItemPointer const &p1, ItemPointer const &p2) const {
+    return (p1.block == p2.block) && (p1.offset == p2.offset);
+  }
+
   ItemPointerComparator(const ItemPointerComparator &) {}
   ItemPointerComparator() {}
 };
 
 struct ItemPointerHasher {
   size_t operator()(const ItemPointer &item) const {
-    return std::hash<oid_t>()(item.block) ^ std::hash<oid_t>()(item.offset);
+    // This constant is found in the CityHash code
+    // [Source libcuckoo/default_hasher.hh]
+    // std::hash returns the same number for unsigned int which causes
+    // too many collisions in the Cuckoohash leading to too many collisions
+    return (std::hash<oid_t>()(item.block)*0x9ddfea08eb382d69ULL) ^
```

Comment:
Can you expand a little on this observation? Item pointers are already unique in a table, why would we need to randomize them further? Are block ID's shared across tables? Does this only affect multi-table txn's? Do you have collision rate #'s? If you want to strengthen the hash, should we just an actually strong hash like murmur or city?

Reply:
So std::hash does a one to one mapping, essentially returning the same uint32_t. ItemPointers are unique across all tables since it is a combination of block(`tile_group_id`) and offset(`tuple_id`). The `tile_group_id` is unique in a database. Once cuckoohash receives the 4 byte hash value, it uses the lower bytes for the default bucket and higher bytes for the  alternate bucket. Since most blocks in the test cases are less than 0x10000, the higher bytes always map to the alternate hash value. As the tuple_ids are non-unique, the lower bytes also have collisions. 

This hash function used is actually the same as city hash.  We could switch to Murmur hash but then cuckoohash uses city hash and it doesn't seem to cause too many collisions.